src/mlpack/methods/reinforcement_learning/sac_impl.hpp

```diff
@@ -0,0 +1,343 @@
+/**
+ * @file methods/reinforcement_learning/sac_impl.hpp
+ * @author Nishant Kumar
+ *
+ * This file is the implementation of SAC class, which implements the
+ * soft actor-critic algorithm.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_RL_SAC_IMPL_HPP
+#define MLPACK_METHODS_RL_SAC_IMPL_HPP
+
+#include <mlpack/prereqs.hpp>
+
+#include "sac.hpp"
+
+namespace mlpack {
+namespace rl {
+
+template <
+  typename EnvironmentType,
+  typename QNetworkType,
+  typename PolicyNetworkType,
+  typename UpdaterType,
+  typename ReplayType
+>
+SAC<
+  EnvironmentType,
+  QNetworkType,
+  PolicyNetworkType,
+  UpdaterType,
+  ReplayType
+>::SAC(TrainingConfig& config,
+       QNetworkType& learningQ1Network,
+       PolicyNetworkType& policyNetwork,
+       ReplayType& replayMethod,
+       UpdaterType qNetworkUpdater,
+       UpdaterType policyNetworkUpdater,
+       EnvironmentType environment):
+  config(config),
+  learningQ1Network(learningQ1Network),
+  policyNetwork(policyNetwork),
+  replayMethod(replayMethod),
+  qNetworkUpdater(std::move(qNetworkUpdater)),
+  #if ENS_VERSION_MAJOR >= 2
+  qNetworkUpdatePolicy(NULL),
+  #endif
+  policyNetworkUpdater(std::move(policyNetworkUpdater)),
+  #if ENS_VERSION_MAJOR >= 2
+  policyNetworkUpdatePolicy(NULL),
+  #endif
+  environment(std::move(environment)),
+  totalSteps(0),
+  deterministic(false)
+{
+  // Set up q-learning and policy networks.
+  learningQ2Network = learningQ1Network;
+  targetQ1Network = learningQ1Network;
+  targetQ2Network = learningQ2Network;
+  if (learningQ1Network.Parameters().is_empty())
+    learningQ1Network.ResetParameters();
+  if (learningQ2Network.Parameters().is_empty())
+    learningQ2Network.ResetParameters();
+  if (targetQ1Network.Parameters().is_empty())
+    targetQ1Network.ResetParameters();
+  if (targetQ2Network.Parameters().is_empty())
+    targetQ2Network.ResetParameters();
+  if (policyNetwork.Parameters().is_empty())
+    policyNetwork.ResetParameters();
```

Comment:
Let's remove the if conditions here and Reset verytime, since we already know the copy constructor will not copy the weights.

Reply:
Yeah, I'll keep this in mind in the next commit